### DISCUSSION TITLE
Executing command to open dialog from ViewModel

### DISCUSSION THREAD

**Discussion by billlhead** [AvaloniaUI/Avalonia#6778] [Q&A]
Hello,
I am writing a desktop app and I need to have a modal dialog for the user to change settings.

Following the example of the Music Store App tutorial, I have a SettingsVIew and SettingsViewModel and a command on the MainWindowViewModel which successfully opens the dialog
```
public class MainWindowViewModel : ViewModelBase
    {
        public Interaction<SettingsViewModel, Unit> ShowSettingsDialog { get; }
        public ICommand ShowSettingsCommand { get; }

        public MainWindowViewModel()
            {
                ShowSettingsDialog = new Interaction<SettingsViewModel, Unit>();

                ShowSettingsCommand = ReactiveCommand.CreateFromTask(async () =>
                {
                    var settings = new SettingsViewModel();
                    await ShowSettingsDialog.Handle(settings);
                });
            }
    }
```
Then in the MainWindow view, I have this code-behind:
```
public partial class MainWindow : ReactiveWindow<MainWindowViewModel>
    {
        public MainWindow()
        {
            ...
            this.WhenActivated(d => d(ViewModel!.ShowSettingsDialog.RegisterHandler(DoShowSettingsDialogAsync)));
        }

        private async Task DoShowSettingsDialogAsync(InteractionContext<SettingsViewModel, Unit> interaction)
        {
            var dialog = new SettingsWindow();
            dialog.DataContext = interaction.Input;
            
            var result = await dialog.ShowDialog<Unit>(this);
            interaction.SetOutput(result);
        }
    }
```

I have a toolbar button's Command bound to ShowSettingsCommand and it opens the dialog like it should, but I also need to open the dialog on start if some settings are blank.

If I try to run `ShowSettingsCommand.Execute(null)` from the MainWindowViewModel constructor nothing happens, but if I run with Rider's debugger attached, it stops at the following block in RxApp.cs
```
 MainThreadScheduler.Schedule(() =>
                {
#pragma warning disable CA1065 // Avoid exceptions in constructors -- In scheduler.
                    throw new UnhandledErrorException(
                        "An object implementing IHandleObservableErrors (often a ReactiveCommand or ObservableAsPropertyHelper) has errored, thereby breaking its observable pipeline. To prevent this, ensure the pipeline does not error, or Subscribe to the ThrownExceptions property of the object in question to handle the erroneous case.",
                        ex);
#pragma warning restore CA1065
                });	
```

What is the correct way to open this dialog from the View Model's constructor? 
Is there a simpler way to open the dialog if I don't need to worry about returning any data from the dialog? 
It just sets some values in a static instance of a Settings model.

Thank you

**Comment by karloyz** [comment]
I encounter the same issue, how is your issue solved? Thanks

**Reply by thevortexcloud** [reply]
If you are using ReactiveUI you can instead do it when the view model is activated by implementing `IActivatableViewModel`. Which will be fired whenever the view is ready to use the view model.

https://www.reactiveui.net/docs/handbook/when-activated

One of the main problems with trying to do things in the constructor is the view model is likely not associated with a view at that point.
